docs/WitnessMatchingManifesto.md

```diff
@@ -0,0 +1,273 @@
+# Protocol Witness Matching Manifesto
+
+Author: [Suyash Srijan](https://www.github.com/theblixguy) 
+
+### Jargon
+
+Before we dive into the document, let's clarify some jargon:
+
+This document uses some jargon, which some readers may not be aware of:
+
+- A _protocol requirement_ (or just _requirement_) is a declaration inside a protocol that all conforming types must satisfy.
+- A _protocol witness_ (or just _witness_) is a value or a type that satisfies a protocol requirement.
+
+For example:
+
+```swift
+protocol P {
+  func foo() // This is a protocol requirement
+}
+
+struct S: P {
+  func foo() { // This is a protocol witness
+    ...
+  }
+}
+```
+
+## Introduction
+
+In Swift (at the time of this writing), protocol witnesses have to pretty much match exactly with a protocol requirement. For example:
+
+```swift
+protocol P {
+  func foo()
+}
+
+struct S1: P {
+  func foo() { ... } // This matches
+}
+
+struct S2: P {
+  func foo(arg: Int = 0) { ... } // This is not a match
+}
+```
+
+```swift
+protocol Q {
+  func getView() -> UIView
+}
+
+struct S1: Q {
+  func getView() -> UIView { ... } // This matches
+}
+
+struct S2: Q {
+  func getView() -> MyCustomUIViewSubclass {  // This is not a match 
+    ...
+  }
+}
+```
+
+However, there are some exceptions to these rules, for example:
+
+1. A non-throwing function can witness a throwing function requirement:
+
+```swift
+protocol P {
+  func foo() throws
+}
+
+struct S: P {
+  // Even though 'foo' here is not marked 'throws', it is
+  // still able to satisfy the protocol requirement. The
+  // caller has to handle the possibility that it might 
+  // throw, but the witness is ofcourse free to throw or 
+  // not to, depending on whether it makes sense in the
+  // implementation.
+  func foo() { ... }
+}
+```
+
+2. A non-failable initializer can witness a failable initializer requirement:
+
+```swift
+protocol P {
+  init?()
+}
+
+struct S: P {
+  // Similarly, the initializer here is free to return
+  // a newly created value or just return 'nil' instead.
+  init() {}
+}
+```
+
+3. A non-escaping closure can witness one that is marked `@escaping`:
+
+```swift
+protocol P {
+  func foo(completion: @escaping() -> ()) 
+}
+
+struct S: P {
+  // Similar to above, you can use a non-escaping closure here
+  // if you want.
+  func foo(completion: () -> ()) {
+    ...
+  }
+}
+```
+
+Some of these rules seem reasonable, but one might argue that the existing model is just too restrictive in general (for example, subtyping is not allowed here, but is allowed in classes). This leads us to ask a very basic question - what kinds of differences are reasonable when matching a protocol witness to a protocol requirement?
```

Comment:
Hmm, the segue here is still pretty awkward.  You just mentioned a bunch of exceptions, so readers will naturally think they're what "some of these rules" refers to, but actually it refers to the *non-exceptions*.  How about this basic flow for the introduction:

>  In Swift, a witness that's declared exactly like a requirement will generally match it.  For example:
>
> ```
> (just some obvious positive examples, no negative examples yet)
> ```
>
> Currently, a few specific differences are allowed:
>
> 1. foo
> 2. bar
> 3. baz
>
> However, almost any other kind of mismatch is forbidden.  For example:
>
> ```
> (here you can have some negative examples, including one that returns a subclass)
> ```
>
> Many people have argued that the existing model is too restrictive and that more mismatched witnesses should be allowed.  It's also been suggested that the existing model feels inconsistent with other parts of the language; for example, the example above where the witness's return type is a subclass of the requirement's return type is allowed with class method overrides.  In order to figure out how respond to these statements, we first have to address a very basic question: what should the model be for how we expect witnesses to match with requirements?

Reply:
That's much better, thanks!